User Preferences page - alternate contact info

bruinc opened the pull request.

bruinc commented on src/assets/src/components/preferences.tsx:
```diff
@@ -0,0 +1,83 @@
+interface PreferencesEditorProps {
+    user: User;
+    disabled: boolean;
+    onUpdateInfo: (phone_number: string) => void;
```
All of the attributes in models.ts are snake case; would you want me to change all of them to camel case?

jlost replied:
Nah, the models should stay how they are - they're reflective of the JSON that is returned by the API, and the API uses snake_case because it's written in Python. (a typescript `interface` just hints to typescript how it expects a data structure to look for type checking purposes)

bruinc commented on src/assets/src/components/preferences.tsx:
```diff
@@ -0,0 +1,83 @@
+    //Setup interactions
+    const updateInfo = async (phone_number: string) => {
+        interactions.next(true);
+        await api.updateMyUser(user_id, phone_number)
```
Would `return await doRefresh()` do that as well?

bruinc commented:
@jlost I think I've addressed all of your comments so far.

jlost commented on src/officehours_api/migrations/0015_profile_phone_number.py:
```diff
@@ -0,0 +1,41 @@
+            model_name='meeting',
+            name='attendees',
+            field=models.ManyToManyField(through='officehours_api.Attendee', to=settings.AUTH_USER_MODEL),
+        ),
```
Any idea why these `AlterField`s were generated or what they're doing? It looks like `attendee.user` and `meeting.attendees` have been changed to link to `Profile` instead of `User`? But I don't see any major changes in `models.py`.

bruinc replied:
I changed them to link to Profile while I was playing around with things, then changed them back, that's why there's four of those. Only the first migration in this file actually does anything. Could I just delete the four AlterFields since they cancel each other out?

jlost commented on src/officehours_api/serializers.py:
```diff
@@ -28,7 +29,24 @@ def get_my_queue(self, obj):
+
+class ProfileSerializer(serializers.ModelSerializer):
+    user = NestedUserSerializer(required=True)
+    #phone_number = serializers.CharField(source='pk')
```
rm

jlost commented on src/officehours_api/nested_serializers.py:
```diff
@@ -1,12 +1,20 @@
+
+    class Meta:
+        model = Profile
+        fields = ['user', 'phone_number'] """
```
rm (you can always find deleted code in your git history later. I see what you were thinking here, and I'm not sure yet which is the best way)

jlost commented on src/assets/src/components/preferences.tsx:
```diff
@@ -0,0 +1,90 @@
+    const updateInfo = async (phoneNumber: string) => {
+         
+        return await api.updateMyUser(userId, phoneNumber)
+        
```
You should remove the extra spacing here

jlost commented on src/assets/src/components/preferences.tsx:
```diff
@@ -0,0 +1,90 @@
+}
+
+function PreferencesEditor(props: PreferencesEditorProps) {
+    var currentPhoneField = ""
```
It looks like you're trying to persist state. Did this come from example code somewhere? It seems to be working when I run it, but I think a bug would manifest pretty quickly because of the (somewhat weird) way React handles state. If you look up the `useState` hook and see the way I'm using it in other components, that's the template that I think you should be following in this spot.

bruinc replied:
This was the first way I thought of keeping track of what the user has typed in the phone number field without having to update the user's phone number with every key stroke. I will definitely look at how you've used the `useState` hook and adapt this.

bruinc commented:
@jlost I resolved your feedback on my code, and now I'm going to work on the usability considerations you mentioned.

bruinc commented:
@jlost Feel free to take a look at what I have so far for the usability considerations. I also have some questions and comments below.

**Form**
- Putting the phone number field inside the form somehow disabled the dropdown for choosing the country code, and I'm not sure why.
- Hitting the enter key in the phone number field does not save/submit the form. I think this might be because I don't have an explicit `<input/>` element, but I'm not sure how to use the input tag with the `react-phone-input-2` component.
- Regarding this default behavior, how should this scale up when there are multiple fields in the form? Would hitting enter in any field submit the whole form?

**Validation**
- The current validation behavior is checking for 11 characters in the phone number upon clicking the save button. If it's true, the phone number is updated, if not, the form is re-loaded, and the field is reset back to what it was at the last save. This validation scheme is not ideal because many country codes are more than one digit, and some countries' area codes are four digits. I'll poke around online to see if I can find any regular expressions or logic for phone number validation.

bruinc commented:
Update: the issue with the country code drop down being disabled has been fixed.

I added basic length validation for USA phone numbers. If the user made a change to the field and the selected country is the USA, then it only updates the phone number if the length is exactly 11. If the user made a change and the selected country is any other country, there is no validation and it submits. 

I also added the case if the user deletes everything in the field, it will update their phone number to be the empty string. I felt like users should be able to change their mind and completely delete their phone number.

@jlost Let me know if there are any changes to this procedure that you'd like me to make. Are there any fields that should be added to the Preferences page before this PR is merged? Or should any additions to the Preferences page beyond the phone number be separate issues?

jlost commented on src/officehours_api/serializers.py:
```diff
@@ -28,7 +29,23 @@ def get_my_queue(self, obj):
+    
+    def update(self, instance, validated_data):
+        print("***SERIALIZER VALIDATED DATA:", validated_data)
+        profile = validated_data.pop('profile')
```
I take it `validated_data` is a dict here? I would typically prefer `validated_data['profile']` (AKA the `__getitem__` accessor) over `.pop(...)` because `pop` mutates the underlying array, which can cause confusion. I do see that `validated_data` is being passed into the `super().update(...)` after the pop, so maybe there's a reason to remove that key that I'm not seeing?

bruinc replied:
My understanding was that the User serializer doesn't need to package the whole profile model with it, it just needs the phone number. So I take the profile out of the validated data, update the user fields with the remaining validated data, then manually update the phone number field.

bruinc replied:
In commit 103543a I changed the `pop()` to just accessing the profile from the validated data, and it seems to work exactly the same.

jlost commented on src/officehours_api/serializers.py:
```diff
@@ -28,7 +29,23 @@ def get_my_queue(self, obj):
         return serializer.data
+    
+    def update(self, instance, validated_data):
+        print("***SERIALIZER VALIDATED DATA:", validated_data)
```
Looks like you left this in by mistake while debugging

jlost commented on src/assets/src/services/api.ts:
```diff
@@ -167,6 +167,20 @@ export const getMyUser = async (user_id: number) => {
+            phone_number: phone_number,
+        }),
+    });
+    console.log(resp)
```
Debugging line to remove

bruinc commented:
I'm going to merge 1.2 into this branch so it can  have the correct migrations. 

@jlost This branch has the new dependency `react-phone-input-2`, so whenever I go to or from this branch, I've been having to run `docker-compose build`. Will we have to do something special when merging into dev to make sure that it installs the new dependency?

jlost commented:
Everything should work fine after we merge into dev, essentially because deployments happen with a `docker build` which is almost identical to the way your local environment gets built.

bruinc commented:
@jlost Merged into dev!
